### PR TITLE
Multi-subnet [DRAFT]

### PR DESCRIPTION
**Pull request by jchen6585**
**What type of PR is this?**

feature

Disregard CRD changes being in v1alpha1, will update to v1alpha2 when it is good to go
**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs from IPAMD/CNI showing the issue**:

**Testing done on this change**:

Manually tested creating an eniconfig with multiple subnets
Integration test suites:
- custom networking: passing
- ipamd: failing (debugging)
- cni: failing (debugging)

**Will this PR introduce any new dependencies?**:

Yes

**Will this break upgrades or downgrades? Has updating a running cluster been tested?**:

**Does this change require updates to the CNI daemonset config files to work?**:

No

**Does this PR introduce any user-facing change?**:

Yes

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/awsutils/awsutils.go [hunk 1]
```diff
@@ -740,8 +741,40 @@ func (cache *EC2InstanceMetadataCache) awsGetFreeDeviceNumber() (int, error) {
...
+		return "", errors.Wrap(err, "AllocENI: unable to describe subnets")
+	}
+
+	sort.SliceStable(result.Subnets, func(i, j int) bool {
+		return *result.Subnets[j].AvailableIpAddressCount < *result.Subnets[i].AvailableIpAddressCount
+	})
+
+	var eniID string
+	for _, element := range result.Subnets {
+		log.Infof("THIS IS THE SUBNET ID: %s", *element.SubnetId)
```
Comment (jchen6585) [inline pkg/awsutils/awsutils.go:766]: these are just some debug statements when i was manually testing, will remove

File: charts/aws-vpc-cni/crds/customresourcedefinition.yaml [hunk 2]
```diff
@@ -8,12 +8,49 @@ spec:
...
+                    type: string
+                  type: array
+                subnet:
+                  type: string
+                subnets:
+                  items:
+                    type: string
+                  type: array
+              type: object
+              oneOf:
```
Comment (jdn5126) [inline charts/aws-vpc-cni/crds/customresourcedefinition.yaml:41]: Nice find!

File: pkg/awsutils/awsutils.go [hunk 3]
```diff
@@ -740,8 +741,40 @@ func (cache *EC2InstanceMetadataCache) awsGetFreeDeviceNumber() (int, error) {
 
 // AllocENI creates an ENI and attaches it to the instance
 // returns: newly created ENI ID
-func (cache *EC2InstanceMetadataCache) AllocENI(useCustomCfg bool, sg []*string, subnet string, numIPs int) (string, error) {
-	eniID, err := cache.createENI(useCustomCfg, sg, subnet, numIPs)
+func (cache *EC2InstanceMetadataCache) AllocENI(useCustomCfg bool, sg []*string, subnets []*string, numIPs int) (string, error) {
+	describeSubnetInput := &ec2.DescribeSubnetsInput{
```
Comment (jdn5126) [inline pkg/awsutils/awsutils.go:745]: This should only be done for custom networking, as in the normal case, we do not want to make a Describe call

File: pkg/awsutils/awsutils.go [hunk 4]
```diff
@@ -740,8 +741,40 @@ func (cache *EC2InstanceMetadataCache) awsGetFreeDeviceNumber() (int, error) {
...
+	ec2ApiReq.WithLabelValues("DescribeSubnets").Inc()
+	awsAPILatency.WithLabelValues("ModifyNetworkInterfaceAttribute", fmt.Sprint(err != nil), awsReqStatus(err)).Observe(msSince(start))
+	if err != nil {
+		checkAPIErrorAndBroadcastEvent(err, "ec2:DescribeSubnets")
+		awsAPIErrInc("DescribeSubnets", err)
+		ec2ApiErr.WithLabelValues("DescribeSubnets").Inc()
+		return "", errors.Wrap(err, "AllocENI: unable to describe subnets")
+	}
+
+	sort.SliceStable(result.Subnets, func(i, j int) bool {
```
Comment (jdn5126) [inline pkg/awsutils/awsutils.go:760]: nit: Can you add a comment here that we sort subnets by available IP address count before determining the subnet to use?

File: pkg/ipamd/ipamd.go [hunk 5]
```diff
@@ -876,17 +876,27 @@ func (c *IPAMContext) tryAllocateENI(ctx context.Context) error {
 			return err
 		}
 
-		log.Infof("ipamd: using custom network config: %v, %s", eniCfg.SecurityGroups, eniCfg.Subnet)
+		log.Infof("ipamd: using custom network config: %v, %s", eniCfg.SecurityGroups, eniCfg.Subnets)
```
Comment (jdn5126) [inline pkg/ipamd/ipamd.go:879]: nit: Let's move this log to below when we find which field to use

**Review by jdn5126 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**